### PR TITLE
Micrometer compatibility mode for Metrics

### LINKED ISSUE #6544: Micrometer compatibility mode for Metrics
Users migrating from Spring Boot and Micrometer might want to expose the same JVM metrics as they used to, so that they don't have to rework their Grafana dashboards. 
The naming and organization of base/vendor metrics exposed by Quarkus is different than what Micrometer does, so it would be useful to have a configuration option that changes them to align with that scheme. In particular, we should mimic the way how `jvm` metrics are handled in https://github.com/micrometer-metrics/micrometer/tree/master/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jvm so we will have to add some new metrics that we don't normally have, and change metadata of some metrics that we do normally have

### PR DESCRIPTION
**Pull request by jmartisk**
Fixes #6544

### CODE REVIEW

**Comment by geoand** [comment]
@ebullient I know you were interested in this :)

**Comment by ebullient** [comment]
I am, indeed. I'll likely bash on the boundaries of this one some more for my DevNexus talk.. :-P Thanks, Jan!

**Comment by gsmet** [comment]
@jmartisk could you ping the original user asking for that so that he gives it a try?

**Comment by jmartisk** [comment]
@gsmet Sure, I asked in the mailing list. Let's see.
For the record I tested it against this dashboard https://grafana.com/grafana/dashboards/3308 (this dashboard uses a bit more metrics than just `jvm` ones, so far this PR contains just `jvm` metrics, we could potentially add a few more if there is enough interest)

**Comment by wedamartinez** [comment]
@jmartisk This is Alejandro, thank you very much for your work, to you and your team!  It looks awesome. I tested and I have almost all that the jvm metrics that we had with micrometer.

**Comment by wedamartinez** [comment]
@jmartisk I have a question if I have this parameter activated, I will be able to add customized "application" metrics? for the moment we don't have any Quarkus microservice using this feature but for sure we are going to use that soon. I am going to create a POC to test that use case.

**Comment by ebullient** [comment]
@wedamartinez .. I am on that path as well, as I have a talk that uses micrometer's custom metrics, and I would like those to be usable w/o change. So we'll see how that goes. ;)

**Comment by wedamartinez** [comment]
Cool also works! thanks!

# HELP application_com_weekendesk_activities_activitytype_http_ActivityTypeResource_performedChecks_total How many primality checks have been performed.
# TYPE application_com_weekendesk_activities_activitytype_http_ActivityTypeResource_performedChecks_total counter
application_com_weekendesk_activities_activitytype_http_ActivityTypeResource_performedChecks_total 5.0

File: extensions/smallrye-metrics/deployment/src/main/java/io/quarkus/smallrye/metrics/deployment/SmallRyeMetricsProcessor.java [hunk 1]
```diff
@@ -103,6 +103,15 @@
         @ConfigItem(name = "extensions.enabled", defaultValue = "true")
         public boolean extensionsEnabled;
 
+        /**
+         * Apply Micrometer compatibility mode, where instead of regular 'base' and 'vendor' metrics,
+         * Quarkus exposes the same 'jvm' metrics that Micrometer does. Application metrics are unaffected by this mode.
```
Comment (wedamartinez) [inline extensions/smallrye-metrics/deployment/src/main/java/io/quarkus/smallrye/metrics/deployment/SmallRyeMetricsProcessor.java]: It's seems that they are affected

File: extensions/smallrye-metrics/deployment/src/main/java/io/quarkus/smallrye/metrics/deployment/SmallRyeMetricsProcessor.java [hunk 2]
```diff
@@ -103,6 +103,15 @@
...
+         * The use case is to facilitate migration from Micrometer-based metrics, because original dashboards for JVM metrics
+         * will continue working without having to rewrite them.
```
Comment (wedamartinez) [inline extensions/smallrye-metrics/deployment/src/main/java/io/quarkus/smallrye/metrics/deployment/SmallRyeMetricsProcessor.java]: thanks 👍

File: extensions/smallrye-metrics/deployment/src/main/java/io/quarkus/smallrye/metrics/deployment/SmallRyeMetricsProcessor.java (hunk 1 above)
Comment (jmartisk) [reply extensions/smallrye-metrics/deployment/src/main/java/io/quarkus/smallrye/metrics/deployment/SmallRyeMetricsProcessor.java]: I don't know what you mean here. Is it about the `application_` prefix being there? That's the way it is normally (without the compatibility mode)

**Comment by jmartisk** [comment]
> @jmartisk I have a question if I have this property activated, I will be able to add customized "application" metrics? for the moment we don't have any Quarkus microservice using any custom metric but for sure we are going to use that soon. I am going to create a POC to test that use case.

@wedamartinez Well that's currently out of scope, that would be a tough one. Micrometer uses a completely different programming API (annotations etc.) than Quarkus (which is based on MicroProfile Metrics), has different metric types, different strategy for exporting the metrics, etc. So what you're asking about sounds more like we would need to introduce a completely new alternative extension that supports Micrometer metrics. I can't quite imagine supporting the Micrometer API and somehow automatically 'bridging' all of it into MicroProfile Metrics without turning it into a complete mess.
The scope of this enhancement is to at least map the basic (built-in) metrics. Application-specific metrics will have to be rewritten manually to use the MP Metrics API.

**Comment by wedamartinez** [comment]
> > @jmartisk I have a question if I have this property activated, I will be able to add customized "application" metrics? for the moment we don't have any Quarkus microservice using any custom metric but for sure we are going to use that soon. I am going to create a POC to test that use case.
> 
> @wedamartinez Well that's currently out of scope, that would be a tough one. Micrometer uses a completely different programming API (annotations etc.) than Quarkus (which is based on MicroProfile Metrics), has different metric types, different strategy for exporting the metrics, etc. So what you're asking about sounds more like we would need to introduce a completely new alternative extension that supports Micrometer metrics. I can't quite imagine supporting the Micrometer API and somehow automatically 'bridging' all of it into MicroProfile Metrics without turning it into a complete mess.
> The scope of this enhancement is to at least map the basic (built-in) metrics. Application-specific metrics will have to be rewritten manually to use the MP Metrics API.

Thank you very much!, I completely agree.

**Comment by jmartisk** [comment]
To be more clear, my previous comment was to say that it's not supported to use an application that includes application metrics through Micrometer API without rewriting them to MP Metrics. The ability to use MP Metrics API is not affected if this mode is active.

@gsmet @geoand It looks to me this is ready, do you guys have any concerns?

File: extensions/smallrye-metrics/runtime/src/main/java/io/quarkus/smallrye/metrics/runtime/SmallRyeMetricsRecorder.java [hunk 3]
```diff
@@ -583,4 +604,193 @@ private void memoryPoolMetrics(MetricRegistry registry, List<String> names) {
         }
     }
 
+    private void micrometerJvmGcMetrics(MetricRegistry registry, List<String> names, ShutdownContext shutdown) {
+        if (!ImageInfo.inImageCode()) {
```
Comment (geoand) [inline extensions/smallrye-metrics/runtime/src/main/java/io/quarkus/smallrye/metrics/runtime/SmallRyeMetricsRecorder.java]: I was just about to ask how the code knows to not expose metrics in GraalVM mode and stumbled upon this before I could :).

:+1:

**Review by geoand — APPROVED** [review]
I don't know a ton about metrics, but this seems reasonable to me

**Comment by gastaldi** [comment]
Works great! Thanks!

**Comment by wedamartinez** [comment]
Thanks! god job!

**Pull request merged**